src/equihash.rs

```diff
@@ -0,0 +1,471 @@
+    state.update(nonce);
+
+    let mut rows = Vec::new();
+    for i in indices {
+        let hash = generate_hash(&state, i / p.indices_per_hash_output());
+        let start = ((i % p.indices_per_hash_output()) * p.n / 8) as usize;
+        let end = start + (p.n as usize) / 8;
+        rows.push(Node {
+            hash: expand_array(&hash.as_bytes()[start..end], p.collision_bit_length(), 0),
+            indices: vec![*i],
+        });
+    }
+
+    let mut hash_len = p.hash_length();
+    while rows.len() > 1 {
+        let mut cur_rows = Vec::new();
+        for pair in rows.chunks(2) {
+            let a = &pair[0];
+            let b = &pair[1];
+            if !has_collision(a, b, p.collision_byte_length()) {
+                // error!("Invalid solution: invalid collision length between StepRows");
+                return false;
+            }
+            if b.indices_before(a) {
+                // error!("Invalid solution: Index tree incorrectly ordered");
+                return false;
+            }
+            if !distinct_indices(a, b) {
+                // error!("Invalid solution: duplicate indices");
+                return false;
+            }
+            cur_rows.push(Node::from_children_ref(a, b, p.collision_byte_length()));
+        }
+        rows = cur_rows;
+        hash_len -= p.collision_byte_length();
+    }
+
+    assert!(rows.len() == 1);
+    return rows[0].is_zero(hash_len);
+}
+
+fn tree_validator(p: &Params, state: &Blake2b, indices: &[u32]) -> Option<Node> {
+    if indices.len() > 1 {
+        let end = indices.len();
+        let mid = end / 2;
+        match tree_validator(p, state, &indices[0..mid]) {
+            Some(a) => match tree_validator(p, state, &indices[mid..end]) {
+                Some(b) => {
+                    if !has_collision(&a, &b, p.collision_byte_length()) {
+                        // error!("Invalid solution: invalid collision length between StepRows");
+                        return None;
+                    }
+                    if b.indices_before(&a) {
+                        // error!("Invalid solution: Index tree incorrectly ordered");
+                        return None;
+                    }
+                    if !distinct_indices(&a, &b) {
+                        // error!("Invalid solution: duplicate indices");
+                        return None;
+                    }
+                    Some(Node::from_children(a, b, p.collision_byte_length()))
+                }
+                None => None,
+            },
+            None => None,
+        }
+    } else {
+        let i = indices[0];
+        let hash = generate_hash(&state, i / p.indices_per_hash_output());
+        let start = ((i % p.indices_per_hash_output()) * p.n / 8) as usize;
+        let end = start + (p.n as usize) / 8;
+        Some(Node {
+            hash: expand_array(&hash.as_bytes()[start..end], p.collision_bit_length(), 0),
+            indices: vec![i],
+        })
```

Comment:
It would be nice to extract this else block in to a function. This would make more clear the similarities and differences between this function and the iterative version.